NotificationHubs/src/com/windowsazure/messaging/HttpClientManager.java

```diff
@@ -1,33 +1,86 @@
 
 import org.apache.http.impl.nio.client.CloseableHttpAsyncClient;
-import org.apache.http.impl.nio.client.HttpAsyncClients;
+import org.apache.http.impl.nio.client.HttpAsyncClientBuilder;
+import org.apache.http.client.config.RequestConfig;
 
 public class HttpClientManager {
-	private static CloseableHttpAsyncClient httpAsyncClient;
-	
-	public static CloseableHttpAsyncClient getHttpAsyncClient() {
-		if(httpAsyncClient == null) {
-			synchronized(HttpClientManager.class) {
-				if(httpAsyncClient == null) {
-					CloseableHttpAsyncClient client = HttpAsyncClients.createDefault();
-					client.start();
-					httpAsyncClient = client;	    	   
-				}
-			}
-		}
-		  
-		return httpAsyncClient;
-	}
-		
-	public static void setHttpAsyncClient(CloseableHttpAsyncClient httpAsyncClient) {
-		synchronized(HttpClientManager.class) {
-			if(HttpClientManager.httpAsyncClient == null) {
-				HttpClientManager.httpAsyncClient = httpAsyncClient;
-			}
-			else{
-				throw new RuntimeException("HttpAsyncClient was already set before or default one is being used.");
-			}
-		}
-	}
+
+    private static CloseableHttpAsyncClient httpAsyncClient;
+
+    // A timeout value of zero is interpreted as an infinite timeout.
+    // A negative value is interpreted as undefined (system default).
+    // https://hc.apache.org/httpcomponents-client-ga/httpclient/apidocs/org/apache/http/client/config/RequestConfig.html#getConnectionRequestTimeout()
+
+    // The timeout in milliseconds used when requesting a connection from the connection manager.
+    private static int connectionRequestTimeout = -1;
+
+    // The timeout in milliseconds until a connection is established.
+    private static int connectionTimeout = -1;
+
+    // The socket timeout in milliseconds, which is the timeout for waiting for data or,
+    // put differently, a maximum period inactivity between two consecutive data packets.
+    private static int socketTimeout = -1;
+
+    public static CloseableHttpAsyncClient initializeHttpAsyncClient() {
+        synchronized (HttpClientManager.class) {
+            if (httpAsyncClient == null) {
+                RequestConfig config = RequestConfig.custom()
+                        .setConnectionRequestTimeout(connectionRequestTimeout)
+                        .setConnectTimeout(connectionTimeout)
+                        .setSocketTimeout(socketTimeout)
+                        .build();
+                CloseableHttpAsyncClient client = HttpAsyncClientBuilder.create()
+                        .setDefaultRequestConfig(config)
+                        .build();
+                client.start();
+                httpAsyncClient = client;
+            } else {
+                throw new RuntimeException("initializeHttpAsyncClient cannot be called twice or after setHttpAsyncClient.");
+            }
+        }
+        return httpAsyncClient;
+    }
+
+    public static CloseableHttpAsyncClient getHttpAsyncClient() {
+        return httpAsyncClient;
```

Comment:
The method `getHttpAsyncClient()` is called directly by `NotificationHub` and `NamespaceManager`. With this change users will need to make code changes to explicitly call `initializeHttpAsyncClient`, otherwise they will get NullRef exception. We should check if `httpAsyncClient` is null and if so call `initializeHttpAsyncClient`